File: Task3A_B/task3a_b/src/main/java/com/cps2004/RPG.java
package com.cps2004;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.node.ObjectNode;

public class RPG extends VideoGame{
    private String story;
    private boolean multiplayer;

    // Default constructor
    public RPG(){
        super();
        this.story = "";
        this.multiplayer = false;
    }

    // Parameterized constructor
    public RPG(String story, boolean multiplayer, String name, int id, Supplier supplier, int price, int volume, String platform, String genre){
        super(name, id, supplier, price, volume, platform, genre);
        this.story = story;
        this.multiplayer = multiplayer;
    }

    // Setters and getters
    public String getStory(){ return this.story;}
    public boolean getMultiplayer(){ return this.multiplayer;}
    public void setStory(String story){ this.story = story;}
    public void setMultiplayer(boolean multiplayer){ this.multiplayer = multiplayer;}

    // Display product information
    public void display_product_info(){
        super.display_product_info();
        System.out.println("Product story: " + story);
        System.out.println("Product multiplayer: " + multiplayer);
    }

    @Override
    // Calculate discounted price
    public double calculate_discount(String month, int quantity){
        if ((month.equals("January") || month.equals("February") || month.equals("March")) && quantity >= 20) {
            return getPrice() * 0.2;
        } else {
            return getPrice() * 0.1;
        }
    }

}

File: Task1B/rpg.cpp
#include "rpg.h"

// Default constructor
RPG::RPG() {
    this->story = "";
    this->multiplayer = false;
}

// Parameterised constructor
RPG::RPG(string name, int price, Supplier* supplier, int volume, int productID, string platform, string developer, string story, bool multiplayer) : VideoGame(name, price, supplier, volume, productID, platform, developer) {
    this->story = story;
    this->multiplayer = multiplayer;
}

// Getters and setters
string RPG::getStory() {
    return this->story;
}

bool RPG::getMultiplayer() {
    return this->multiplayer;
}

void RPG::setStory(string story) {
    this->story = story;
}

void RPG::setMultiplayer(bool multiplayer) {
    this->multiplayer = multiplayer;
}

// Display product information
void RPG::display_product_info() {
    VideoGame::display_product_info();
    cout << "Story: " << this->story << endl;
    cout << "Multiplayer: " << this->multiplayer << endl;
}

// Calculate discount
double RPG::calculate_discount(string month, int quantity) {
    if ((month == "January" || month == "February" || month == "March") && quantity >= 20) {
        return getPrice() * 0.2;
    } else {
        return getPrice() * 0.1;
    }
}

// Save product information
void RPG::save(ostream& out) const {
    out << "rpg" << "\n";
    VideoGame::save(out);
    out << story << "," << multiplayer << "\n";
}

// Load product information
void RPG::load(istream& in) {
    VideoGame::load(in);
    
    string line;
    getline(in, line, '\n');
    stringstream ss(line);

    string token;
    getline(ss, token, ',');
    story = token;

    getline(ss, token, ',');
    multiplayer = stoi(token);
}